File: Java/treehopper/src/main/java/io/treehopper/TreehopperUsb.java
package io.treehopper;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import io.treehopper.interfaces.Connection;
import io.treehopper.interfaces.I2c;
import io.treehopper.interfaces.Pwm;

/**
 * Created by jay on 12/27/2015.
 */

public class TreehopperUsb {

    public static Settings Settings = new Settings();

    private static final Logger logger = LogManager.getLogger("TreehopperUsb");

    public Pin[] pins = new Pin[20];
    private Connection usbConnection;
    private boolean led;
    private boolean connected;

    Object comsLock = new Object();

    public TreehopperUsb(Connection connection) {
        usbConnection = connection;
        usbConnection.setPinReportListener(this);
        for (int i = 0; i < 20; i++)
            pins[i] = new Pin(this, i);

        pwm1 = new HardwarePwm(pins[7]);
        pwm2 = new HardwarePwm(pins[8]);
        pwm3 = new HardwarePwm(pins[9]);

        i2c = new HardwareI2c(this);
        spi = new Spi(this);
        hardwarePwmManager = new HardwarePwmManager(this);
    }

    public I2c i2c;
    public Spi spi;
    public Pwm pwm1;
    public Pwm pwm2;
    public Pwm pwm3;
    public HardwarePwmManager hardwarePwmManager;

    public boolean getConnected() {
        return connected;
    }

    public boolean connect() {
        if(connected) return true;
        connected = true;
        return usbConnection.open();
    }

    public void disconnect() {
        if(!connected) return;
        connected = false;
        usbConnection.close();
    }

    public String getName() {
        return usbConnection.getName();
    }

    public String getSerialNumber() {
        return usbConnection.getSerialNumber();
    }

    public boolean isLed() {
        return led;
    }

    public void setLed(boolean led) {
        this.led = led;
        byte[] DataToSend = new byte[2];
        DataToSend[0] = (byte) DeviceCommands.LedConfig.ordinal();
        DataToSend[1] = (byte) (led ? 0x01 : 0x00); // Unicode 16-bit strings are 2 bytes per character
        usbConnection.sendDataPeripheralChannel(DataToSend);
    }

    public Connection getConnection() {
        return usbConnection;
    }


    public void onPinReportReceived(byte[] pinReport) {
        if (pinReport[0] != 0x00) {
            int i = 1;
            for (Pin pin : pins) {
                pin.UpdateValue(pinReport[i++], pinReport[i++]);
            }
        }
    }

    public void sendPeripheralConfigPacket(byte[] dataToSend) {
        usbConnection.sendDataPeripheralChannel(dataToSend);
    }

    public byte[] receiveCommsResponsePacket(int numBytesToRead) {
        return usbConnection.readPeripheralResponsePacket(numBytesToRead);
    }
}

File: C++/API/src/TreehopperUsb.cpp
#include "TreehopperUsb.h"
#include "HardwarePwm.h"
#include "HardwareI2c.h"

namespace Treehopper 
{
	TreehopperUsb::TreehopperUsb(UsbConnection& connection) :
		connection(connection),
		pwmManager(*this),
		i2c(*this),
		spi(*this),
		pwm1(*this, 7),
		pwm2(*this, 8),
		pwm3(*this, 9)
	{
		for (int i = 0; i < numberOfPins; i++)
			pins.emplace_back(this, i);
	}

	TreehopperUsb::~TreehopperUsb()
	{
		if (_isDestroyed)
			return;

		cout << "Destructor" << endl;
        disconnect();
        //delete &connection;
		_isDestroyed = true;
	}

	bool TreehopperUsb::connect()
	{
		if (!connection.open())
		{
			return false;
		}

		isConnected = true;
		pinListenerThread = thread(&TreehopperUsb::pinStateListener, this);

		reinitialize();
		return true;
	}

	void TreehopperUsb::disconnect()
	{
		if (!isConnected) return;

		reinitialize(); // leave the board where we found it

		isConnected = false;
        if(pinListenerThread.joinable())
            pinListenerThread.join(); // block this thread until the listener exits
		connection.close();
	}

	void TreehopperUsb::reinitialize()
	{
		uint8_t data[2];
		data[0] = (uint8_t)DeviceCommands::ConfigureDevice;
		sendPeripheralConfigPacket(data, 2);
	}

	wstring TreehopperUsb::serialNumber()
	{
		return connection.serialNumber();
	}

	wstring TreehopperUsb::name()
	{
		return connection.name();
	}

	void TreehopperUsb::led(bool value)
	{
        if(!isConnected) return;
        
        _led = value;
		uint8_t data[2];
		data[0] = (uint8_t)DeviceCommands::LedConfig;
		data[1] = _led;
		connection.sendDataPeripheralChannel(data, 2);
	}

	bool TreehopperUsb::led()
	{
		return _led;
	}

	void TreehopperUsb::sendPinConfigPacket(uint8_t* data, size_t len)
	{
        if(!isConnected) return;
		connection.sendDataPinConfigChannel(data, len);
	}

	void TreehopperUsb::sendPeripheralConfigPacket(uint8_t* data, size_t len)
	{
        if(!isConnected) return;
		connection.sendDataPeripheralChannel(data, len);
	}

	void TreehopperUsb::receivePeripheralConfigPacket(uint8_t * data, size_t numBytesToRead)
	{
        if(!isConnected) return;
		connection.receiveDataPeripheralChannel(data, numBytesToRead);
	}

	void TreehopperUsb::pinStateListener()
	{
		while (isConnected)
		{
                if(!connection.receivePinReportPacket(buffer)) continue;
                
                for (int i = 0; i < numberOfPins; i++)
                    pins[i].updateValue(buffer[i * 2 + 1], buffer[i * 2 + 2]);
        }
	}
}